### DISCUSSION TITLE
Remove LGPL v3 dependency (unibilium)

### DISCUSSION THREAD

**Discussion by setjmp** [neovim/neovim#31966] [Ideas]
unibilium is licensed as LGPL 3, and such a license prevents certain types of deployment of neovim at large corporations which take license compliance very seriously.  I came to write this feature request not out of whim, but as a response to a hurdle that I and others are facing.

I have not done a thorough review of all the third party dependencies, but AFAIK this is the only LGPL 3 blocker.  

Replacing unibilium will increase adoption of neovim tremendously.

**Comment by jamessan** [comment]
Only the _new_ code developed for nvim has been submitted under the Apache license.  All of the original Vim code is still licensed under its GPL-like (or more likely, according to II.2.e., GPL-2+) license.

Given that, I find it hard to believe that unibilium's LGPL-3 license is the limiting factor here, if your reviewers are doing their due diligence.

**Comment by jamessan** [comment]
If you _really_ can't have unibilium, then you can choose [not to use the TUI](https://github.com/neovim/neovim/blob/eada8f5aaae0c072571c87b6dbd3c7992541d698/CMakeLists.txt#L418-L421), but then you'll need some other UI to use.

**Comment by setjmp** [comment]
GPL v2 has a very different hurdle for approval than GPL v3, because it is a different license with different terms and conditions, and GPLv3 is absolutely a limiting factor in this case.  Your comment suggests the people who are setting or enforcing policy may be uninformed or sloppy.  I am not at all offended by this suggestion as it seems like a natural hypothesis to propose given all the general incompetence in the world.  However, please accept my explanation that this is not the case; the people who make these decisions are qualified and principled.

Also, I do not know of any (L)GPL v3 code in the dependencies of vim 8.x which I took a quick glance at.  Notably, I find no reference to unibilium in that code base.

**Comment by justinmk** [comment]
>  I do not know of any (L)GPL v3 code in the dependencies of vim 8.x

Vim 8.1 inlines xdiff, which is LGPL. But not v3.

I vaguely recall questioning unibilium's license years ago, and IIRC @tarruda said the author was OK with our inclusion of it.

Although, I'm rather alarmed by this, I must have forgotten that it was v3.

**Comment by mqudsi** [comment]
I feel like (and forgive me if I'm mistaken) OP might be confusing LGPLv3 with GPLv3. It's a different beast entirely, and if a company is fine with GPLv2 it's hard to believe they would have a problem with LGPLv3.

**Comment by setjmp** [comment]
mqudsi@ There is no confusion on my part.  You are right it is different, and it is similar as it states right at the top: "This version of the GNU Lesser General Public License incorporates the terms and conditions of version 3 of the GNU General Public License, supplemented by the additional permissions listed below."  If the LGPL v3 license were C source code, it would be dominated by lines of code caused by #including GPL v3.

**Comment by throwaway1037** [comment]
Free software projects should stand firm with strong copyleft to combat the threat of freedom being stripped away by middlemen along the chain of distribution.

**Comment by throwaway1037** [comment]
Please don't hide the message of freedom. It weakens our community.

**Comment by eestrada** [comment]
>  ...new code developed for nvim has been submitted under the Apache license.

> I vaguely recall questioning unibilium's license years ago, and IIRC @tarruda said the author was OK with our inclusion of it.

The easiest path forward is probably just relicensing historical pre-fork code. 

[There are only 5 contributors to the original code before the fork](https://github.com/mauke/unibilium/graphs/contributors). Those are the only people who need to consent, since all new code is already Apache licensed, from what it sounds like. But even if the new post-fork code also needs to relicense, [there are only 14 contributors post-fork](https://github.com/neovim/unibilium/graphs/contributors).

For context: in a different project I contributed to years ago, [they chose to relicense](https://github.com/racket/racket/issues/1570) and just replying to a Github issue with a templated affirmation was considered sufficient. That had hundreds of contributors and still succeeded in relicensing. Should be possible here with the much smaller number of contributors, and easier than replacing a core dependency.

Although it sounds like [there is a plan to move to a different toolkit (?)](https://github.com/neovim/neovim/pull/21832#discussion_r1071143289), asking permission to relicense is still probably a faster way forward for the time being until that happens.

**Comment by justinmk** [comment]
> [There are only 5 contributors to the original code before the fork](https://github.com/mauke/unibilium/graphs/contributors). Those are the only people who need to consent, since all new code is already Apache licensed, from what it sounds like

Glad to see @mauke is alive, was silent for a few years. I'm feeling optimistic he will (1) reply to this message and (2) relicense unibilium 🤞

**Comment by mauke** [comment]
Why would I want to relicense unibilium?

**Comment by eestrada** [comment]
> Why would I want to relicense unibilium?

The [most active fork of Unibilium](https://github.com/mauke/unibilium/forks) is the one maintained by the Neovim community. No other fork (including the original Unibilium repo) has seen any activity in the last two year. This indicates that the Neovim has a vested interested in continuing development of the codebase.

Re-licensing will allow active development to continue in the Neovim community. However, maintaining the current licensing will likely cause Unibilium to be dropped from Neovim at a future date, because it decreases adoption of Neovim in corporate environments. Regardless of whether those corporations are right or wrong in their hesitance to allow GPLv3/LGPLv3 licensed code, the reality is that it causes issues for end users who want to use Neovim at their job, but cannot due to a licensing conflict.

As the OP says:

> Replacing unibilium will increase adoption of neovim tremendously.

If re-licensing Unibilium is a viable option, there are benefits to all parties:

1. Unibilium continues to receive bug fixes and updates via the Neovim community.
2. Unibilium sees increased adoption via being included in Neovim.
3. Neovim community avoids doing the work required to replace a core dependency.
4. Neovim sees increased adoption via licensing that can be used in more places.

As I see it, the potential paths forward for Neovim are the following (ordered by level of difficulty from least to most difficult):

1. Receive consent to re-license Unibilium.
2. Re-write/refactor Neovim to use a different library than Unibilium.
3. Convince every corporate lawyer at every major tech company that GPLv3/LGPLv3 licensing is actually OK and they should have nothing to worry about.

The last one is obviously a little bit tongue-in-cheek; if you have ever worked directly with corporate lawyers it is even funnier. For as much of a win as number 3 would be, it isn't feasible in a timeline that can be easily estimated, whereas the first two are.

That is why I brought up the possibility of re-licensing: it is the simplest and fastest way to resolve this issue for Neovim, and there are benefits to Unibilium like continuing development and adoption.

@mauke If you don't want to re-license after reading the rationale I've given, that is OK too, but at least the question has been asked and the option explored.

**Comment by clason** [comment]
> However, maintaining the current licensing will likely cause Unibilium to be dropped from Neovim at a future date, because it decreases adoption of Neovim in corporate environments.

I don't know where you got that idea from. We have _zero_ intention of doing that.

**Comment by eestrada** [comment]
> I don't know where you got that idea from.

Based on the fact that this issue to remove the Unibilium dependency remains open, and [this comment](https://github.com/neovim/neovim/pull/21832#discussion_r1071143289) from @justinmk which links to this issue which says what it will be replaced with:

> ...that issue is parked until we can replace unibilium with https://github.com/leonerd/libtickit

Perhaps that comment is outdated based on the current direction the core team is headed. I only know what is publicly available. It wasn't my intention to speak for anyone else.

>  We have zero intention of doing that.

Leaving this issue open as it is now makes it seem like there is some non-zero amount of intention of doing that. If there is no intention to remove the dependency, I suggest this issue be closed with a comment to indicate that intention, or that this issue be converted into a discussion to indicate that is in, in fact, just a discussion and not an issue that is meant to be addressed. 

A closed issue can always be re-opened later if people's minds change. Leaving it open makes it seem like it is meant to be addressed at some point.

**Comment by clason** [comment]
You're reading too much into this. We were looking at libtickit, with an eye towards switching to it (for technical reasons!), but no firm decision or concrete plans so far. And it was certainly not for the reason you state.

> A closed issue can always be re-opened later if people's minds change. Leaving it open makes it seem like it is meant to be addressed at some point.

You clearly haven't been around here much...

**Comment by clason** [comment]
Anyway, converted to discussion, since that is a thing now (and wasn't when the issue was opened).

**Comment by eestrada** [comment]
> You're reading too much into this.

Seems like it. This issue came up at work and I thought to offer the alternative of re-licensing. Sorry for jumping to conclusions regarding future intentions on the dependency. I should have caveated my initial response. 

> You clearly haven't been around here much...

Correct :)

> Anyway, converted to discussion, since that is a thing now (and wasn't when the issue was opened).

Mitchell Hashimoto makes [a good post here](https://x.com/mitchellh/status/1852039089547309552) about ensuring that all GH issues are actionable. That was with a smaller private beta, so maybe not viable here on a public project, but seems like a good paradigm when it is possible. With some automation, converting stale issues to discussions could help clean up the issue backlog to only planned or actionable features/bugs/refactors.

**Reply by clason** [reply]
Yes, not every project is the same. Ghostty is in a _very_ different place from Neovim.

(The relicensing suggestion was apposite and reasonable; I just wanted to correct some incorrect assumptions in your "recap".)

**Reply by justinmk** [reply]
> With some automation, converting stale issues to discussions could help clean up the issue backlog to only planned or actionable features/bugs/refactors

Please *don't* do this. 

And I wish this wasn't converted to a discussion, because it can't exactly be reverted. This is an open issue and I 100% want to replace unibilium with something that can match our license.

**Reply by eestrada** [reply]
> And I wish this wasn't converted to a discussion, because it can't exactly be reverted.

I apologize for suggesting it.

> Please don't do this.

Issue automation has at least as many problems as benefits. I can now see why automatic discussion/issue conversion would cause problems; it seems it is not a two way door. I was (incorrectly) under the impression that it was trivially reversible when I suggested converting this to a discussion.

**Comment by bfredl** [comment]
> Neovim sees increased adoption via licensing that can be used in more places. [..] Convince every corporate lawyer at every major tech company that GPLv3/LGPLv3 licensing is actually OK and they should have nothing to worry about.

This is a hypothetical stated as a fact. IF there actually is big business companies wanting to use neovim but the _loicensing_ is getting in the way, they can come talk to us any time.

**Comment by setjmp** [comment]
Original poster here.  Thank you for following up.

I have since switched companies and no longer have the context.  Also, this issue is no longer important to me personally.  IIRC, though:

The license did play a role in limiting the means in which I was able to provide the tool to others.  There were other options proposed though.  I don't believe I followed up on those; I just gave up.  You can keep that account in a basket of anecdotes.

Having worked at two of the FAANGs, I can confirm/re-affirm that GPL3 is avoided as a matter of policy at both places.  Whether it has a role in slowing the adoption of Neovim as it currently is or finding new innovative use cases for bundling/embedding Neovim with other systems is hard, if not impossible, to quantify.

I will say that classical vim is unencumbered by this license, or at least it was last time I checked.

If you don't resolve this issue I am sure neovim will continue to be a successful project.  If you are interested in maximizing the potential for adoption of neovim and have no qualms with the idea that business plays a role in that growth... I believe it is worth resolving this issue.

**Reply by clason** [reply]
Thank you for the update. The fact is simply that supporting corporate environments is very low on our list of priorities: there's no philosophical objection; there's simply (much) better uses of our very limited resources as far as increasing the attractiveness of Neovim as a text editor is concerned.

And as bfredl said; corporations _do_ have the resources if they feel sufficiently strongly about it. (Some companies in fact support OSS developments, but those who do towards Neovim never raised this issue. I suspect that there's a strong anti-correlation here...)

**Reply by justinmk** [reply]
> I will say that classical vim is unencumbered by this license, or at least it was last time I checked.

Yup. Can confirm that this is policy at my current employer.

**Reply by eestrada** [reply]
> Having worked at two of the FAANGs, I can confirm/re-affirm that GPL3 is avoided as a matter of policy at both places.

Can also confirm.

Have worked at BigTech/FAANG company and at a large-ish corporation (so it is not just BigTech/FAANG issue) where they both had/have blanket policies regarding open source licenses, GPLv3 being on the "Avoid/Not allowed" list. Can't remember policy regarding LGPLv3.

> corporations do have the resources if they feel sufficiently strongly about it.

It is not so much the corporations who suffer (or care) about not having Neovim available because of license conflict. It is the employees working in the corporations who are constrained by policy that they have little possibility of changing. That is how I read the original issue, and what spurred me to suggest an alternative to removal of the dependency.

**Comment by justinmk** [comment]
Closed by https://github.com/neovim/neovim/pull/36381 and related PRs. To build without Unibilium, set `ENABLE_UNIBILIUM=off` during cmake configuration.
